src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs

```diff
@@ -0,0 +1,455 @@
+            }
+        }
+
+        public virtual int KeyExchangeStrength
+        {
+            get
+            {
+                return _SslState.KeyExchangeStrength;
+            }
+        }
+
+        //
+        // Stream contract implementation.
+        //
+        public override bool CanSeek
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        public override bool CanRead
+        {
+            get
+            {
+                return _SslState.IsAuthenticated && InnerStream.CanRead;
+            }
+        }
+
+        public override bool CanTimeout
+        {
+            get
+            {
+                return InnerStream.CanTimeout;
+            }
+        }
+
+        public override bool CanWrite
+        {
+            get
+            {
+                return _SslState.IsAuthenticated && InnerStream.CanWrite;
+            }
+        }
+
+        public override int ReadTimeout
+        {
+            get
+            {
+                return InnerStream.ReadTimeout;
+            }
+            set
+            {
+                InnerStream.ReadTimeout = value;
+            }
+        }
+
+        public override int WriteTimeout
+        {
+            get
+            {
+                return InnerStream.WriteTimeout;
+            }
+            set
+            {
+                InnerStream.WriteTimeout = value;
+            }
+        }
+
+        public override long Length
+        {
+            get
+            {
+                return InnerStream.Length;
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                return InnerStream.Position;
+            }
+            set
+            {
+                throw new NotSupportedException(SR.net_noseek);
+            }
+        }
+
+        public override void SetLength(long value)
+        {
+            InnerStream.SetLength(value);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException(SR.net_noseek);
+        }
+
+        public override void Flush()
+        {
+            _SslState.Flush();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                _SslState.Close();
+            }
+            finally
+            {
+                base.Dispose(disposing);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return _SslState.SecureStream.Read(buffer, offset, count);
+        }
+
+        public void Write(byte[] buffer)
+        {
+            _SslState.SecureStream.Write(buffer, 0, buffer.Length);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _SslState.SecureStream.Write(buffer, offset, count);
+        }
```

Comment:
Is it possible the inner stream overrides ReadAsync/WriteAsync/FlushAsync?  If so, this stream should ideally override those as well and delegate to the inner counterparts.